[dtensor] add placement strategy and einsum strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98227

This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

wanchaol committed:
[dtensor] add placement strategy and einsum strategy

This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

XilunWu commented on test/distributed/_tensor/test_basic_strategy.py:
```diff
@@ -0,0 +1,120 @@
+        input_dims, output_dim = EinsumDims.parse_equation(equation)
+        edims = EinsumDims.parse_dims(input_dims, output_dim)
+
+        self.assertEqual(len(edims.batch_dims), 0)
+        self.assertEqual(len(edims.contracting_dims), 1)
+        self.assertEqual(len(edims.lhs_free_dims), 1)
+        self.assertEqual(len(edims.rhs_free_dims), 1)
```
would it be better to compare on the value itself instead of its length?
```
self.assertEqual(edims.batch_dims, [])
self.assertEqual(edims.contracting_dims, ['k'])
```

wanchaol replied:
sure, updated

XilunWu commented on test/distributed/_tensor/test_basic_strategy.py:
```diff
@@ -0,0 +1,120 @@
+        edims = EinsumDims.parse_dims(input_dims, output_dim)
+
+        self.assertEqual(len(edims.batch_dims), 1)
+        self.assertEqual(len(edims.contracting_dims), 0)
```
isn't `b` a contracting dim?

XilunWu commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,173 @@
+)
+
+
+@dataclass
```
thx for extracting the common logic into a class!

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,173 @@
+
+    # generate strategies for each mesh dim
+    for mesh_dim in range(mesh.ndim):
+        if mesh.size(mesh_dim) <= 1:
```
then in this case `len(all_mesh_dim_strategies) != mesh.ndim`, right? would it be better to keep them equal?

XilunWu commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,176 @@
+    all_mesh_dim_strategies = []
+
+    # generate strategies for each mesh dim
+    for mesh_dim in range(mesh.ndim):
+        if mesh.size(mesh_dim) <= 1:
+            # no strategy for mesh dim with size 1
+            # TODO: maybe see if submesh needs this and we want to fully replicate
+            continue
```
then in this case `len(all_mesh_dim_strategies) != mesh.ndim`, right? would it be better to keep them equal?

XilunWu reviewed (approved):
lgtm

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

zdevito commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,176 @@
+class EinsumDims:
+    contracting_dims: List[str]
+    batch_dims: List[str]
+    lhs_free_dims: List[str]
```
These are the dims that appear in the lhs and output but not the rhs. `free` feels like confusing name because it suggests "free variable" meaning it's not bound in the lhs. Maybe `lhs_only` or `lhs_output_only`.

wanchaol replied:
make sense, changed.

zdevito commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,176 @@
+        for lhs_dim in edims.lhs_free_dims:
+            lhs_free_dim = output_dim.index(lhs_dim)
+            lhs_placement_list: List[Placement] = [
+                Shard(lhs_free_dim),
```
What do the 3 elements of the list correspond to?

wanchaol replied:
it corresponds to output_spec, lhs_input_spec, rhs_input_spec, updated with comments.

zdevito commented on torch/distributed/_tensor/ops/basic_strategy.py:
```diff
@@ -0,0 +1,176 @@
+    # generate strategies for entire mesh
+    strategy_combs = itertools.product(*all_mesh_dim_strategies)
+
+    # TODO: filter out invalid strategies
```
kinda curious what the constraints are, and if there is way to generate the strategies directly based on what would be valid.

wanchaol replied:
The constraints mostly:
1. we need to see if there's invalid strategies base on tensor shapes (I think if we want to shard a tensor dim we need to ensure it's bigger than mesh dim size)
2. maybe there's a tensor dim shard on multiple mesh dims, we need to check if it's still bigger than the mesh size, and see if we want to disable that sometimes.

zdevito reviewed (commented):
Setup looks good. It will be good to see how the strategies get applied to understand more about what the data structures mean and to check if the generated strategies are correct.

wanchaol committed:
Update on "[dtensor] add placement strategy and einsum strategy"


This adds placement strategy to the op schema and implement einsum
strategy. It's the basic building piece for compile mode expansion
and new op implementation

[ghstack-poisoned]

wanchaol commented:
> Setup looks good. It will be good to see how the strategies get applied to understand more about what the data structures mean and to check if the generated strategies are correct.

Thanks @zdevito! Comments addressed. It's actually a bit hard to check if the generated strategies are correct in tests (as there're 16 or 25 possibilities generated), so I assertted the strategy lengths, and printed the results to eyeball verify the correctness. I plan to apply/use this strategy in follow up PRs and try to figure out a way to ensure correctness in tests :)

wanchaol commented:
@pytorchbot merge
